src/grt/src/fastroute/src/FastRoute.cpp

```diff
@@ -1227,4 +1262,239 @@ const char* netName(FrNet* net)
+
+      if (is3DVisualization_) 
+        lastL = gridsL[0];
+      
+      for (int i = 1; i <= routeLen; i++) {
+        const int xreal = w_tile_ * (gridsX[i] + 0.5) + x_corner_;
+        const int yreal = h_tile_ * (gridsY[i] + 0.5) + y_corner_;
+
+        if (is3DVisualization_) {
+          drawLineObject(lastX, lastY, lastL + 1, xreal, yreal, gridsL[i] + 1, painter);
+          lastL = gridsL[i];
+        }
+        else {
+          drawLineObject(lastX, lastY, -1, xreal, yreal, -1, painter);  // -1 to 2D Trees
+        }
+        lastX = xreal;
+        lastY = yreal;
+      }
+    }
+  }
+}
+void FastRouteRenderer::drawCircleObjects(gui::Painter& painter){
+
+  for (int i = 0; i < num_pins_; i++) {
+    const int xreal = w_tile_ * (pinX_[i] + 0.5) + x_corner_;
+    const int yreal = h_tile_ * (pinY_[i] + 0.5) + y_corner_;
+    
+    odb::dbTechLayer* layer = tech_->findRoutingLayer(pinL_[i] + 1);
+    painter.setPen(layer);
+    painter.setBrush(layer);
+    painter.setPenWidth(700);
+    painter.drawCircle(xreal, yreal,1500);
+  }
+}
+
+void FastRouteRenderer::drawObjects(gui::Painter& painter)
+{
+  if (treeStructure_ == 1) {
+  
+    const int deg = stree_.deg;
+    for (int i = 0; i < 2 * deg - 2; i++) {
+      const int x1 = w_tile_ * (stree_.branch[i].x + 0.5) + x_corner_;
+      const int y1 = h_tile_ * (stree_.branch[i].y + 0.5) + y_corner_;
+      const int n = stree_.branch[i].n;
+      const int x2 = w_tile_ * (stree_.branch[n].x + 0.5) + x_corner_;
+      const int y2 = h_tile_ * (stree_.branch[n].y + 0.5) + y_corner_;
+      const int len = abs(x1-x2)+abs(y1-y2);
+      if (len > 0) {
+          painter.setPen(painter.white);
+          painter.setBrush(painter.white);
+          painter.setPenWidth(700);
+          painter.drawLine(x1,y1,x2,y2);
+      }
+    }
+    
+    drawCircleObjects(painter);
+  }
+  else if (treeStructure_ == 2) {
+    
+    drawTreeEdges(painter);
+    
+    drawCircleObjects(painter);
+  }
+}
+
+////////////////////////////////////////////////////////////////
+void FastRouteCore::createDebug()
+{
+  debug_ = new DebugSetting();
+}
+void FastRouteCore::setDebugOn(bool isOn)
+{
+  debug_->isOn_ = isOn;
+}
+void FastRouteCore::setDebugSteinerTree(bool steinerTree)
+{
+  debug_->steinerTree_ = steinerTree;
+}
+void FastRouteCore::setDebugTree2D(bool tree2D)
+{
+  debug_->tree2D_ = tree2D;
+}
+void FastRouteCore::setDebugTree3D(bool tree3D)
+{
+  debug_->tree3D_ = tree3D;
+}
+void FastRouteCore::setDebugNet(const odb::dbNet *net)
+{
+  debug_->net_ = net;
+}
+void FastRouteCore::setDebugRectilinearSTree(bool rectiliniarSTree)
+{
+  debug_->rectilinearSTree_ = rectiliniarSTree;
+}
+
+
+void FastRouteCore::steinerTreeVisualization(stt::Tree& stree, FrNet* net){
+  //init FastRouteRender
+  if (gui_) {
+    if (fastrouteRender_ == nullptr) {
+      fastrouteRender_ = new FastRouteRenderer(this, db_->getTech(), w_tile_, h_tile_, x_corner_, y_corner_);
+      fastrouteRender_->highlight(net);
```

Comment:
does this belong in the initialization?  What happens on a second call to this method?